cql3/functions/functions.cc

```diff
@@ -298,6 +302,82 @@ static shared_ptr<function> get_dynamic_aggregate(const function_name &name, con
     return {};
 }
 
+static
+shared_ptr<function>
+get_set_intersection_function(data_dictionary::database db,
+        const sstring& keyspace,
+        const function_name& name,
+        const std::vector<shared_ptr<assignment_testable>>& provided_args,
+        const sstring& receiver_ks,
+        std::optional<const std::string_view> receiver_cf,
+        const column_specification* receiver) {
+    if (provided_args.size() != 2) {
+        throw exceptions::invalid_request_exception("set_intersection() accepts 2 arguments only");
+    }
+    auto known_arg_types = provided_args
+            | std::views::filter([] (const shared_ptr<assignment_testable>& arg) { return arg->assignment_testable_type_opt().has_value(); })
+            | std::views::transform([] (const shared_ptr<assignment_testable>& arg) -> data_type { return *arg->assignment_testable_type_opt(); })
+            | std::ranges::to<std::vector>();
+
+    if (known_arg_types.empty()) {
+        throw exceptions::invalid_request_exception("set_intersection() can only be called if at least one argument type is known");
+    }
+
+    auto known_set_types = known_arg_types
+        | std::views::transform([] (const data_type& arg_type) {
+            return dynamic_pointer_cast<const set_type_impl>(arg_type);
+        })
+        | std::ranges::to<std::vector>();
+
+    if (!std::ranges::all_of(known_set_types, [] (data_type t) { return t != nullptr; })) {
+        throw exceptions::invalid_request_exception("set_intersection() can only be called if all arguments are of set type");
+    }
+
+    // Normalize everything to be a frozen set
+    for (auto& set_type : known_set_types) {
+        set_type = dynamic_pointer_cast<const set_type_impl>(set_type->freeze());
+    }
+
+    auto unique_remove = std::ranges::unique(known_set_types);
+    known_set_types.erase(unique_remove.begin(), unique_remove.end());
+
+    if (known_set_types.size() != 1) {
+        throw exceptions::invalid_request_exception(fmt::format("set_intersection() can only be called if all arguments are of the same set type: {}",
+                known_set_types | std::views::transform(&abstract_type::name)));
+    }
+
+    auto set_type = known_set_types.front();
+    auto element_type = set_type->get_elements_type();
+
+    return make_native_scalar_function<true>("set_intersection", set_type, {set_type, set_type},
```

Comment:
Maybe it was a mistake to include the word "scalar" in the name of make_native_scalar_function()? After all, this returns a set, not a scalar?
(or maybe "scalar" in this context means "not aggregation", not "not collection"? I don't see any comments in scalar_function.hh or native_scalar_function.hh hinting what these things are for).

Reply:
Yes, scalar is "not aggregation", not "not collection".